tensorflow/core/kernels/resource_variable_ops.cc

```diff
@@ -869,6 +871,99 @@ bool ValidateInput<Variant>(const Tensor& updates) {
   return true;
 }
 
+template <typename Device, typename T, typename Index, scatter_op::UpdateOp op>
+Status DoScatter(OpKernelContext* c, Tensor* params, const Tensor& indices,
+                 const Tensor& updates, Index num_indices);
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+template <typename T>
+Status CopyTensorToHost(OpKernelContext* c, const Tensor& device_tensor,
+                        Tensor* host_tensor) {
+  AllocatorAttributes alloc_attr;
+  alloc_attr.set_on_host(true);
+  alloc_attr.set_gpu_compatible(true);
+  auto stream = c->op_device_context()->stream();
+  TF_RETURN_IF_ERROR(c->allocate_temp(
+      device_tensor.dtype(), device_tensor.shape(), host_tensor, alloc_attr));
+  se::DeviceMemoryBase device_ptr(
+      const_cast<Tensor&>(device_tensor).flat<T>().data(),
+      device_tensor.flat<T>().size() * sizeof(T));
+  stream->ThenMemcpy(host_tensor->flat<T>().data(), device_ptr,
+                     device_tensor.NumElements() * sizeof(T));
+  if (!stream) {
+    return errors::Internal("Failed to copy indices to host");
+  }
+  return Status::OK();
+}
+
+// Copies inputs to the CPU, runs DoScatter on the CPU, then copies output
+// back to GPU. This is useful because the CPU implementation is deterministic
+// and the GPU implementation is not. Tensor inputs to this function must be on
+// the GPU.
+template <typename T, typename Index, scatter_op::UpdateOp Op>
+Status DoScatterOnCpu(OpKernelContext* c, Tensor* params, const Tensor& indices,
+                      const Tensor& updates, Index num_indices) {
+  auto stream = c->op_device_context()->stream();
+
+  Tensor host_indices;
+  TF_RETURN_IF_ERROR(CopyTensorToHost<Index>(c, indices, &host_indices));
+  Tensor host_updates;
+  TF_RETURN_IF_ERROR(CopyTensorToHost<T>(c, updates, &host_updates));
+  Tensor host_params;
+  TF_RETURN_IF_ERROR(CopyTensorToHost<T>(c, *params, &host_params));
+
+  TF_RETURN_IF_ERROR(stream->BlockHostUntilDone());
+  TF_RETURN_IF_ERROR(DoScatter<CPUDevice, T, Index, Op>(
+      c, &host_params, host_indices, host_updates, num_indices));
+
+  // Copy 'host_params' to device.
+  se::DeviceMemoryBase params_ptr(params->flat<T>().data(),
+                                  params->flat<T>().size() * sizeof(T));
+  stream->ThenMemcpy(&params_ptr, host_params.flat<T>().data(),
+                     host_params.NumElements() * sizeof(T));
+  if (!stream) {
+    return errors::Internal("Failed to copy params to device");
+  }
+  // Block host, since 'host_params' cannot be destructed until the copy is
+  // done.
+  TF_RETURN_IF_ERROR(stream->BlockHostUntilDone());
```

Comment:
Can you use `ThenRunAfterNextBlockHostUntilDone` here?

Reply:
That might wait a long time to deallocate `host_params`. Instead I replace `BlockHostUntilDone` with `EventMgr::ThenExecute`.